feat(core): Save memory at execution

cr0hn opened the pull request:
### Context

This PR has a new feature that significantly reduces the memory footprint when Prowler is running.

### Description

This PR aims to reduce Prowler's memory footprint by switching from using the regular Python list and dictionaries to using SQLite3 as a database.

In the first place, it could sound a bit weird, but it's actually a good idea to reduce Prowler's memory footprint, and the performance impact is really low.

In some tests I've run, Prowler's memory footprint went from 5GB to 400-500MB, which is a huge improvement.

In this PR not all checks of all providers are implemented, but only these ones:

- AWS

And for the following AWS services:

- CloudWatch
- S3
- EC2
- ECS
- CloudWatch Logs
- SageMaker
- Lambda
- SSM

In addition, this PR also includes a quick fix that aims to use Prowler with local AWS emulators (e.g., localstack or moto).

### License

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

cr0hn committed:
Enable custom STS endpoint for local debugging

Add support for using a custom STS endpoint when the PROWLER_LOCAL_DEBUG environment variable is set to "1". This allows developers to specify a different AWS endpoint URL for local testing and debugging purposes.

cr0hn committed:
Add SQLite-based persistence layer

Introduce SQLiteDict and SQLiteList classes for data persistence using SQLite. Add mklist and mkdict factory functions to manage different database connections based on environment variables.

Add SQLiteDict and SQLiteList classes for data persistence using SQLite. Add mklist and mkdict factory functions to manage different database connections based on environment variables.

cr0hn committed:
Add force_gc context manager to manage garbage collection

Introduce force_gc to temporarily disable and enable garbage collection within a code block. This utility can be used to better control memory management by ensuring garbage collection is only performed at specific points.

This is useful when you want to ensure that a certain piece of code does not cause memory leaks or excessive garbage collection.

cr0hn committed:
Add unit tests for SQLiteDict and SQLiteList classes

Added comprehensive unit tests for SQLiteDict and SQLiteList in the SQLite persistence module. These tests cover basic operations, error handling, and edge cases ensuring the correctness and robustness of the implementations.

This commit also includes a new test file for the SQLite persistence module, which includes tests for the SQLiteDict and SQLiteList classes. The tests are designed to cover a wide range of scenarios and edge cases, including testing for proper handling of different data types, empty and non-empty lists, and various error conditions.

Additionally, the test file includes a new test case for the SQLite persistence module, which tests the behavior of the SQLiteDict and SQLiteList classes when dealing with duplicate keys. This test ensures that the classes handle duplicate keys correctly and do not raise any exceptions or errors.

Overall, these changes are aimed at improving the reliability and robustness of the SQLite persistence module, making it more resilient to various scenarios and edge cases. The new unit tests provide a comprehensive set of tests to ensure the correctness and reliability of the SQLiteDict and SQLiteList classes.

cr0hn committed:
Add end-to-end tests for AWS resource population

Introduced scripts and documentation for running end-to-end tests with LocalStack. This includes `populate.py` for creating AWS resources in LocalStack, `generate_env_file.sh` for generating environment variables, and a README with detailed instructions.

cr0hn committed:
Refactor list and dict initialization with mklist/mkdict

Replaced standard list and dict initializations with mklist and mkdict functions across multiple service modules. This enhances consistency and readability in handling collections throughout the codebase.

cr0hn committed:
Refactor findings output and optimize memory usage

Introduced GeneratedOutputs class and used mklist to manage findings and compliance outputs more efficiently. Incorporated garbage collection to manage memory consumption during output generation for large findings sets.

The idea behind this commit is to make the tool more memory efficient and to optimize the output generation process. The tool now uses a list to store the findings and compliance outputs, and the list is

cr0hn committed:
Refactor: Use `mklist` for initializing `findings`

Updated several S3 bucket check classes to use `mklist` for initializing the `findings` variable. This change promotes consistency and potential improvements in handling list initialization across multiple files.

cr0hn committed:
Refactor findings list initialization to use mklist()

Replaced the initialization of findings with mklist() in various EC2-related checks across multiple files. This change reduces the memory footprint of the tool by removing the need to allocate a large list of findings.

cr0hn committed:
Refactor findings initialization in ECS check

Modified the initialization of the `findings` variable to use `mklist()` for reducing the memory footprint of the program.

cr0hn committed:
Use mklist for consistent list initialization across files

Replaced list initialization with mklist function in three SSM-related modules. This change ensures consistent a reduction of memory footprint

cr0hn committed:
Update findings initialization to use mklist

Refactored findings initialization across multiple AWS Lambda checks to use mklist from prowler.lib.persistence. This reduces the memory footprint of the findings

cr0hn committed:
Refactor: Use mklist for findings initialization

Replaced direct initialization of the findings list with the mklist function across multiple SageMaker service checks. This reduces the memory footprint

cr0hn committed:
Update findings initialization to use `mklist` function

Replaced all instances where findings lists were directly initialized with an empty list with the `mklist` function from `prowler.lib.persistence` across various cloudwatch-related scripts. This change helps to reduce the memory footprint

cr0hn committed:
Add dill package and update Poetry lockfile

Added the `dill` package to `pyproject.toml` for enhanced serialization capabilities. Updated the `poetry.lock` file to reflect this change and ensure consistent environment setup.

jfagoagas commented:
Hello @cr0hn, this is great! We will review as soon as we can, in the meantime could you merge your branch with the `master` branch since it seems that some things changed since the point your branch was created.

Thanks!

jfagoagas committed:
fix: conflicts

jfagoagas commented on prowler/lib/persistence/__init__.py:
```diff
@@ -0,0 +1 @@
+from .manager import *
```
Is this a good practice? All the linters are flagging it and I think it is just to skip one part of the import path, right?

jfagoagas reviewed (changes requested):
Hello @cr0hn, I just did a first pass to all the new business logic to improve the memory management and I expect to do some functional tests later today to verify that everything is running as expected.

Some other colleagues are reviewing this PR too due to their importance. I left some comments/questions because some parts are not clear to me, please review it when you get a chance.

Regarding tests I didn't review it yet but I'm thinking that if we are introducing a new `end_to_end` folder, we will probably need to wrap the rest into a `unit` folder.

By the way, I fixed all the conflicts and I'm applying all the linters/formatters we have configured. Please, next time remember to follow our contribution guidelines here https://docs.prowler.com/projects/prowler-open-source/en/latest/developer-guide/introduction/#contributing-with-your-code-or-fixes-to-prowler

Thanks!

jfagoagas commented:
@cr0hn could you please review the errors raised from our formatters here https://github.com/prowler-cloud/prowler/actions/runs/10773270870/job/29872693660?pr=4892?

cr0hn committed:
Update prowler/lib/persistence/manager.py

Co-authored-by: Pepe Fagoaga <pepe@verica.io>

cr0hn committed:
Update __init__.py

fix linter errors. This is not a bad practice if you also use "__all__" meta var in the file you're importing

cr0hn committed:
Update ec2_instance_public_ip.py

fix - missing mklist import

cr0hn committed:
Update populate.py

fix - linter errors

cr0hn commented:
I just fixed the linter errors @jfagoagas

jfagoagas commented on prowler/lib/utils/gc.py:
```diff
@@ -0,0 +1,35 @@
+
+
+@contextmanager
+def force_gc(disable_gc=False):
```
What is the point of this context manager? Is it intended for testing and benchmarking?

jfagoagas commented on prowler/lib/utils/gc.py:
```diff
@@ -0,0 +1,35 @@
+        gc.collect()
+
+    if disable_gc:
+        gc.enable()
```
Would not be easier to just call explicitly the GC from `main`?

jfagoagas replied:
We are not sure about this, but we will review that later on.

jfagoagas committed:
Merge branch 'master' into memory-save

vicferpoy commented on prowler/lib/persistence/manager.py:
```diff
@@ -0,0 +1,52 @@
+    prowler_db_connection = os.environ.get('PROWLER_DB_CONNECTION', "memory://")
+    prowler_db_cache_size = os.environ.get('PROWLER_DB_CACHE_SIZE')
+
+    if not prowler_db_connection:
+        return list()
+
+    # In-memory. Default.
+    if prowler_db_connection.startswith('memory://'):
+        return list()
```
This can be unified in a single if.

vicferpoy commented on prowler/lib/persistence/manager.py:
```diff
@@ -0,0 +1,52 @@
+        return list()
+
+    # In-memory. Default.
+    if prowler_db_connection.startswith('memory://'):
```
I don't like using `startswith` here, if the env variable remains as string.

jfagoagas replied:
Added a TODO

cr0hn committed:
Update manager.py

imp - Adding default value to DB Cache Size and improving error handling

vicferpoy commented on prowler/lib/persistence/manager.py:
```diff
@@ -0,0 +1,52 @@
+    """
+    Create a new list with the given name and data.
+    """
+    prowler_db_connection = os.environ.get('PROWLER_DB_CONNECTION', "memory://")
```
I'm confused about the usage of this environment variable. Why does it have to be a string? Wouldn't it make more sense to be a flag to either use sqlite or not?

vicferpoy replied:
I can only see this variable usage as a flag, unless I'm missing something. That would make its name not descriptive enough.

jfagoagas replied:
For now, we will leave it as is, looking to have a DB connection string within the environment variable.

cr0hn committed:
Update sqlite.py

fix - remove the DEFAULT_CACHE_SIZE value from environment var because this is the default value and always should be a fixed value

vicferpoy commented on prowler/lib/persistence/sqlite.py:
```diff
@@ -0,0 +1,267 @@
+
+    def __init__(self, cache_size=2000):
+        self._tmp_path = tempfile.NamedTemporaryFile(prefix='prowler-dict-')
+        self.db_name = self._tmp_path.name
+        self.conn = sqlite3.connect(self.db_name)
```
Same concerns as Pepe about creating a database and a connection every time we create a list or dict.

cr0hn committed:
Update prowler/lib/outputs/outputs.py

Co-authored-by: Pepe Fagoaga <pepe@verica.io>

cr0hn committed:
Update prowler/lib/outputs/outputs.py

Co-authored-by: Pepe Fagoaga <pepe@verica.io>

vicferpoy commented on prowler/lib/persistence/sqlite.py:
```diff
@@ -0,0 +1,267 @@
+        self._tmp_path.close()
+
+    def __add__(self, other):
+        if isinstance(other, SQLiteList):
+            new_list = SQLiteList()
+            for item in self:
+                new_list.append(item)
+            for item in other:
+                new_list.append(item)
+            return new_list
+        elif isinstance(other, list):
+            new_list = SQLiteList()
+            for item in self:
+                new_list.append(item)
+            for item in other:
+                new_list.append(item)
+            return new_list
```
```suggestion
        if isinstance(other, (SQLiteList, list)):
            new_list = SQLiteList()
            new_list.extend(self)
            new_list.extend(other)
            return new_list
```

vicferpoy commented on prowler/lib/persistence/sqlite.py:
```diff
@@ -0,0 +1,267 @@
+            raise TypeError("Unsupported operand type(s) for +: 'SQLiteList' and '{}'".format(type(other).__name__))
+
+    def __iadd__(self, other):
+        if isinstance(other, (SQLiteList, list)):
+            for item in other:
+                self.append(item)
+            return self
+        else:
+            raise TypeError("Unsupported operand type(s) for +=: 'SQLiteList' and '{}'".format(type(other).__name__))
```
```suggestion
        if isinstance(other, (SQLiteList, list)):
            self.extend(other)
            return self
        raise TypeError(f"Unsupported operand type(s) for +=: 'SQLiteList' and '{type(other).__name__}'")
```

cr0hn reviewed (commented):
Answer comments

jfagoagas commented on prowler/__main__.py:
```diff
@@ -302,56 +304,74 @@ def prowler():
+    for finding in findings:
+        finding_outputs.append(Finding.generate_output(global_provider, finding))
+
+    generated_outputs = GeneratedOutputs(
```
We will review it soon, for now it is ok.

jfagoagas commented on prowler/lib/persistence/README.md:
```diff
@@ -0,0 +1,104 @@
+
+### Using the non-memory based data structures
+
+When you want to use a non-memory, this is SQLite, you need to set the `PROWLER_DB_CONNECTION` environment variable to `sqlite://`.
```
It is going to be the default behaviour.

jfagoagas commented on prowler/providers/aws/aws_provider.py:
```diff
@@ -1165,6 +1165,10 @@ def create_sts_session(
             )
+
+            if os.environ.get("PROWLER_LOCAL_DEBUG") == "1":
+                sts_endpoint_url = os.environ["AWS_ENDPOINT_URL"]
```
I'll leave this commented for now since it is just intended for testing.

jfagoagas replied:
STS is a key point of everything around the AWS provider, we will see what to do here.

jfagoagas commented on prowler/lib/persistence/README.md:
````diff
@@ -0,0 +1,104 @@
+For example:
+
+```bash
+> export PROWLER_DB_CONNECTION=sqlite://
````
Thanks!

jfagoagas committed:
fix: address comments

jfagoagas committed:
fix: conflict in poetry

jfagoagas committed:
chore: format

cr0hn commented on prowler/lib/persistence/README.md:
````diff
@@ -0,0 +1,104 @@
+> export PROWLER_DB_CACHE_SIZE=1000
+```
+
+The cache size is the number of items that will be stored in memory. The default value is 2000.
````
I'm referring to the number of memory pages used by SQLite. In the example, the 1000 value will store pages in memory up to a maximum of 1000kiB

jfagoagas committed:
fix: conflict in poetry

jfagoagas committed:
fix(sqlite): Use it by default
